vertex-ai-friendly-meals/android/app/src/main/java/com/google/firebase/example/friendlymeals/ui/home/HomeScreen.kt

```diff
@@ -0,0 +1,224 @@
+import androidx.compose.material3.ExperimentalMaterial3Api
+import androidx.compose.material3.Scaffold
+import androidx.compose.material3.Text
+import androidx.compose.material3.TextField
+import androidx.compose.material3.TextFieldDefaults
+import androidx.compose.material3.TopAppBarDefaults
+import androidx.compose.material3.rememberTopAppBarState
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.graphics.asImageBitmap
+import androidx.compose.ui.input.nestedscroll.nestedScroll
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import androidx.hilt.navigation.compose.hiltViewModel
+import androidx.lifecycle.compose.collectAsStateWithLifecycle
+import com.google.firebase.example.friendlymeals.R
+import com.google.firebase.example.friendlymeals.data.model.Recipe
+import com.google.firebase.example.friendlymeals.ui.shared.LoadingIndicator
+import com.google.firebase.example.friendlymeals.ui.theme.DarkFirebaseYellow
+import com.google.firebase.example.friendlymeals.ui.theme.FriendlyMealsTheme
+import com.google.firebase.example.friendlymeals.ui.theme.LightFirebaseYellow
+import com.google.firebase.example.friendlymeals.ui.theme.MediumFirebaseYellow
+import com.halilibo.richtext.commonmark.Markdown
+import com.halilibo.richtext.ui.BasicRichText
+import kotlinx.serialization.Serializable
+
+@Serializable
+object HomeRoute
+
+@Composable
+fun HomeScreen(
+    viewModel: HomeViewModel = hiltViewModel()
+) {
+    val recipe = viewModel.recipe.collectAsStateWithLifecycle()
+    val loading = viewModel.loading.collectAsStateWithLifecycle()
+
+    HomeScreenContent(
+        onGenerateClick = viewModel::generateRecipe,
+        recipe = recipe.value,
+        loading = loading.value
+    )
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun HomeScreenContent(
+    modifier: Modifier = Modifier,
+    onGenerateClick: (String, String) -> Unit,
+    recipe: Recipe?,
+    loading: Boolean
+) {
+    val scrollBehavior = TopAppBarDefaults.pinnedScrollBehavior(rememberTopAppBarState())
+
+    Scaffold(
+        modifier = Modifier.nestedScroll(scrollBehavior.nestedScrollConnection)
+    ) { innerPadding ->
+        Column(
+            modifier = modifier
+            .fillMaxWidth()
+            .padding(innerPadding)
+            .verticalScroll(rememberScrollState())
+        ) {
+            IngredientsBox(onGenerateClick = onGenerateClick)
+
+            Spacer(modifier = Modifier.size(16.dp))
+
+            if (recipe == null && loading) {
```

Comment:
I think this leaves room for error in case we pass a null recipe and forget to set the loading to true.

Should we maybe use an `OR` condition instead?

Reply:
An `OR` would break the following logic:

- `null` recipe && loading is `false`: show prompt UI, *no loading icon*
- `null` recipe && loading is `true`: show prompt UI, *with loading icon*
-  non `null` recipe, regardless of loading value: show prompt UI, show results UI, *no loading icon*

We can simply use `if (loading)` - but we would still rely on setting loading to `true` in the `ViewModel`. However, I think this is natural since it's part of the business logic. What do you think?

Reply:
SGTM

Reply:
Updated to use `loading` only.